iwp_website/controllers/investor_portal.py

```diff
@@ -14,79 +16,54 @@
+        """Wallet of share owned by the connected user."""
         values = self._prepare_portal_layout_values()
         shareline_mgr = request.env['share.line']
 
-        domain = self.shareline_domain
-
-        searchbar_sortings = {
-            # FIXME: Should be ordered by the name of the structure not
-            # his id.
-            # 'structure': {'label': _('Structure'), 'order': 'structure'},
-            'date': {'label': _('Date'), 'order': 'effective_date asc'},
-            # FIXME: total_amount_line is a computed field that cannot
-            # be ordered
-            # 'amount': {'label': _('Amount'), 'order': 'total_amount_line'},
-        }
-
-        # default sortby order
-        if not sortby:
-            sortby = 'date'
-        sort_order = searchbar_sortings[sortby]['order']
-
-        # FIXME: _get_archive_groups do no use sudo
-        # archive_groups = self._get_archive_groups('share.line', domain)
-        # if date_begin and date_end:
-        #     domain += [('effective_date', '>', date_begin),
-        #                ('effective_date', '<=', date_end)]
-
-        # count for pager
-        shareline_count = shareline_mgr.sudo().search_count(domain)
-        # make pager
-        pager = portal_pager(
-            url="/my/wallet",
-            url_args={
-                'date_begin': date_begin,
-                'date_end': date_end,
-                'sortby': sortby
-            },
-            total=shareline_count,
-            page=page,
-            step=self._items_per_page
-        )
-        # search the count to display, according to the pager data
-        sharelines = shareline_mgr.sudo().search(
-            domain,
-            order=sort_order,
-            limit=self._items_per_page,
-            offset=pager['offset']
-        )
-        request.session['my_sharelines_history'] = sharelines.ids[:100]
-
-        # Grand Amount Total
-        grand_total = sum([sl.total_amount_line for sl in sharelines])
+        # Share lines owned by an investor
+        sharelines = shareline_mgr.sudo().search(self.shareline_domain)
+
+        # Data structure
+        # [
+        #    {
+        #         'structure': structure_id,
+        #         'total_amount': sum_of_total_amount_line,
+        #         'lines': recordset('share.line'),
+        #     },
+        #     {
+        #         ...
+        #     },
+        # ]
+        data = []
+        # Create data structure
+        sharelines = sharelines.sorted(key=lambda r: r.structure.name)
+        grouped_sl = groupby(sharelines, lambda r: r.structure)
+        for (structure, shares) in grouped_sl:
+            item = {}
+            item['structure'] = structure
+            item['lines'] = shareline_mgr  # New empty recordset
+            item['total_amount'] = 0
+            for share in shares:
+                item['total_amount'] += share.total_amount_line
+                item['lines'] += share
```

Comment:
quitte à la jouer fonctionnel, ¡KILL ALL THE LOOPS!
```python
item['total_amount'] = sum(shares.mapped('total_amount_line'))
item['lines'] = shares.mapped('share')
```

Reply:
Ici, `shares` est un objet itérable renvoyé par la fonction python `groupby`. Une fois qu'on l'a parcouru, il est considéré comme vide. Ici je ne parcours qu'une seule fois l'objet itérable. Avec des `map`, on parcourt l'objet plusieurs fois. Ce qui ne va pas marcher, sauf astuce.

Aussi `mapped` ne peut pas être exécuté sur l'objet `shares` car ce n'est pas un recordset. Peut-être existe-t-il une fonction Python qui fait de même et permet par ailleurs d'être exécuté plusieurs fois sur le même objet itérable ? Je suis curieux. :)

Reply:
```python
item['lines'] = list(shares)
item['total_amount'] = sum(map(lambda s: s.total_amount_line, item['lines']))
```

Par exemple

Reply:
Ok, merci ! Très instructif. :)